### DISCUSSION TITLE
Declaration macro

### DISCUSSION THREAD

**Discussion by AlomeProg** [P-p-H-d/mlib#122] [Ideas]
Hi, I'm new to c and I'm wondering how this library will behave if I need to declare a type in a header file to be used in several translation units afterwards, something like forwarded data (in this library, as I know all generated types are reference types of sorts)? 
As I understand now, if I make LIST_DEF(name, int) in the header file, it creates the same set of functions in each translation unit, as they are kind of declared as static inline and it can lead to code growth and hence file growth? (I hope I understand correctly how this all works).
If it works like that, why not create a macro that can just generate the declarations and the definitions the user would be required to generate in the .c file? What do you think of the idea? If I'm wrong, please ask me to correct me

**Comment by P-p-H-d** [comment]
As they are "static inline" the declared functions will only consume executable space if they are used in the current translation unit (otherwise they are simply gone).
So, they are not as much as file growth as you assume. Moreover, if you put a container in a header, it will be part of another data structure as an interface. So I assume the user of the header will use the specialized interface of the data structure you have created, rather than the static inline. What really matter is more a generic data structure like string_t which may be used everywhere in your program (so much more convenient than a char* and do the stuff yourself!).

You can use the global macros `M_USE_DECL` in your headers and `M_USE_DEF` on one C file [link](https://github.com/P-p-H-d/mlib?tab=readme-ov-file#m_use_decl) to generate only one instance of the function per executable (See examples [1](https://github.com/P-p-H-d/mlib/blob/master/example/exm-main.c) [2](https://github.com/P-p-H-d/mlib/blob/master/example/exm-lib.c) and [3](https://github.com/P-p-H-d/mlib/blob/master/example/exm-header.h)
This will work on all M\×*LIB functions, including string functions.

On another work, I plan to integrate a new version of a "shared pointer" interface which would allow full encapsulation of data (you will have separate declaration and definition).

**Reply by AlomeProg** [reply]
That's what it's all about! Thanks, I somehow missed the examples, focusing on the Readme, I should be more attentive. Good luck with the development!
